### DISCUSSION TITLE
Traveling salesman for efficient mapping

### DISCUSSION THREAD

**Discussion by tarexo** [streetcomplete/StreetComplete#4288] [Ideas]
I had this idea floating around inside my head and thought I'd share it to see what you think about it. Its quite simple.

To kill some spare time I thought it might be an interesting feature to select a quest (or a set of quests) in a given area and the app calculates the most efficient route to work them through. This would obviously work much better with quest, which are not very frequently asked (in the area), like e.g. regarding park benches or bus stops. 

I'm no UI/UX designer, just how I'd imagine the feature to work:

The user could **access the feature from the burger menu** (similar to overlays). Tapping on the entry would start a small **wizard** (or maybe rather just a card?), which guides the user through some questions to set parameters.

First the user must choose a (or multiple?) **quest type**. Instead of just giving him the whole list, the app could use the quest currently shown in the **proximity of the user** on the map, **ordered by count** (ascending).

Next the user is asked for some more parameters: "Only use lit streets", "Respect wheelchair accessability", "Maximum length (in m or km)", "Maximum radius around the current location (in m or km)", ...

After all of this is done, the route is **drawn onto the map**. Using a small element in the UI or even just a toast could notify the user, what the **estimated duration for the walk** will be. I do not think that it the position of the user should snap to the route. It should only act as a guide, similar in style to the markings the user leaves on the map when moving.

To remove the track from the map the user would again access the feature from the burger menu (again similar to the overlays). To avoid the need of setting up the feature everytime the user wants to use it, **presets** could be saved. Accessing the presets would (again) be similar to the overlays to keep the UI consistent.

Keep in mind that I just thought about it from the usage perspective. I do not know if it would be desireable for more users than just me, if it would fit with the app and if it would be technically possible without too much of an overhead. But I liked the idea and wanted to share it.

Let me hear what you think! 

Edit: I am most often in small cities or the country side. I do not know if the feature would make much sense in big cities like Berlin, Paris or Lisbon, where quest density ist probably much higher.

Edit2: The quests chosen by the user for routing are probably the only ones the user is interested in. Therefore the rest of the quests could be hidden (or just grayed out/less visible?) while the route feature is used.

**Comment by mnalis** [comment]
That idea seems to me **_very complex_**, and outside the scope of StreetComplete.

If I'd had to do it, I'd probably use another routing app (like [OsmAnd](https://en.wikipedia.org/wiki/OsmAnd)) alongside StreetComplete, and use its waypoints feature to create routing for specified transport mode (which might not be the same if you're pedestrian, cyclist etc!). If I needed a list of POIs to visit I'd probably use [Overpass Turbo](https://wiki.openstreetmap.org/wiki/Overpass_turbo) to create it and export it as `.gpx` for OsmAnd use. 

But most probably I'd just use guessing and common sense to visit the quest I want to pay special attention to (I never manage to solve them all in one pass anyway, so on first one I'd just target closest 80% of interesting quests, and on next possible visits I'd try to solve the ones I missed earlier).
